### PR TITLE
Event Stream Stubbing

### PR DESCRIPTION
**Pull request by alextwoods**
Implement Event Stream Stubbing.

Adds support for stubbing for event streams.  Interface:
```
client.stub_responses(
    :event_stream_operation,
    {
        events: [
            Types::ChatEvent.new(message: 'message'),
            {message: Message.new(headers: {}, payload: 'somedata')}
        ], 
        initial_response: {initial_structure: {message: 'data'}}}
    },
    {
        error: Hearth::NetworkingError # an error in initial request, no events are sent
    }
)
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: codegen/projections/rails_json/lib/rails_json/middleware.rb [hunk 1]
```diff
@@ -53,8 +53,10 @@ def self.build(config, options = {})
         stack.use(Middleware::RequestId)
         stack.use(Hearth::Middleware::Send,
           client: config.http_client,
+          response_events: false,
           stub_data_class: Stubs::AllQueryStringTypes,
           stub_error_classes: [],
+          stub_message_encoder: Hearth::EventStream::Binary.const_get(:MessageEncoder).new,
```
Comment (mullermp) [inline codegen/projections/rails_json/lib/rails_json/middleware.rb:59]: Why do you need const_get here?

File: codegen/projections/rails_json/lib/rails_json/middleware.rb [hunk 2]
```diff
@@ -53,8 +53,10 @@ def self.build(config, options = {})
         stack.use(Middleware::RequestId)
         stack.use(Hearth::Middleware::Send,
           client: config.http_client,
+          response_events: false,
```
Comment (mullermp) [inline codegen/projections/rails_json/lib/rails_json/middleware.rb:56]: Is there a true example here?

File: codegen/projections/white_label/spec/event_stream_stubs_spec.rb [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+          Hearth::EventStream::HeaderValue.new(value: v, type: 'string')
+      end
+      Hearth::EventStream::Message.new(
+        headers: headers,
+        payload: StringIO.new('{"message":"event_message"}')
+      )
+    end
+
+    it 'signals the stubbed event' do
+      subject.stub_responses(:start_event_stream, {
```
Comment (mullermp) [inline codegen/projections/white_label/spec/event_stream_stubs_spec.rb:34]: Wonky formatting on all of these.

File: hearth/lib/hearth/middleware/event_streams.rb [hunk 4]
```diff
@@ -90,6 +92,7 @@ def setup_response_events(context)
           event_handler: @event_handler
         )
         context.response.body = decoder
+        context.metadata[:event_handler] = @event_handler
```
Comment (mullermp) [inline hearth/lib/hearth/middleware/event_streams.rb:95]: Should there be a named key for this? Or is metadata fine?

**Review by mullermp — APPROVED** [review]
Nice! I probably missed a bunch, but I trust the test coverage.

File: codegen/projections/rails_json/lib/rails_json/middleware.rb (hunk 1 above)
Comment (alextwoods) [reply codegen/projections/rails_json/lib/rails_json/middleware.rb:59]: This is how the module is intended to work from codegen - similar to the way we do an error module.

The specific encoding of messages is protocol specific (all existing ones currently use this same logic, but its supposed to be configurable in codegen per protocol) - so the protocol itself defines just a module to use and that module is expected to porovide a `MessageEncoder` and a `MessageDecoder`.

File: codegen/projections/rails_json/lib/rails_json/middleware.rb (hunk 2 above)
Comment (alextwoods) [reply codegen/projections/rails_json/lib/rails_json/middleware.rb:56]: Not for rails json - I don't think it really makes sense to try to add. 

But theres good examples in white label with this.

File: codegen/projections/white_label/spec/event_stream_stubs_spec.rb (hunk 3 above)
Comment (alextwoods) [reply codegen/projections/white_label/spec/event_stream_stubs_spec.rb]: ugh, rubocop kept mucking these up.  I think I've fixed them all.

**Pull request merged**